launcher/impersonate.go

```diff
@@ -0,0 +1,159 @@
+	}
+
+	resp, err := f.client.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("HTTP call returned error: %v", err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		fmt.Println(resp)
+		return nil, fmt.Errorf("HTTP call returned non-OK status: %v", resp.Status)
+	}
+
+	defer resp.Body.Close()
+	respBody, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read response body: %v", err)
+	}
+
+	return respBody, nil
+}
+
+func (f *impersonatedTokenFetcher) fetchAccessToken(serviceAccount, authToken string, delegates []string) (string, error) {
+	url := iamCredentialsURL + serviceAccount + accessTokenEndpoint
+
+	reqBody, err := json.Marshal(accessTokenReq{
+		Delegates: delegates,
+		Scopes:    []string{"https://www.googleapis.com/auth/cloud-platform"},
+	})
+	if err != nil {
+		return "", fmt.Errorf("unable to marshal request body: %v", err)
+	}
+
+	respBody, err := f.postHTTP(url, authToken, reqBody)
+	if err != nil {
+		return "", fmt.Errorf("token retrieval failed: %v", err)
+	}
+
+	var response accessTokenResp
+	if err = json.Unmarshal(respBody, &response); err != nil {
+		return "", fmt.Errorf("unable to unmarshal response: %v", err)
+	}
+
+	return response.AccessToken, nil
+}
+
+func (f *impersonatedTokenFetcher) fetchIDToken(serviceAccount, authToken, audience string) (string, error) {
+	url := iamCredentialsURL + serviceAccount + idTokenEndpoint
+
+	reqBody, err := json.Marshal(idTokenReq{Audience: audience, IncludeEmail: true})
+	if err != nil {
+		return "", fmt.Errorf("unable to marshal request body: %v", err)
+	}
+
+	respBody, err := f.postHTTP(url, authToken, reqBody)
+	if err != nil {
+		return "", fmt.Errorf("token retrieval failed: %v", err)
+	}
+
+	var response idTokenResp
+	if err = json.Unmarshal(respBody, &response); err != nil {
+		return "", fmt.Errorf("unable to unmarshal response: %v", err)
+	}
+
+	return response.Token, nil
+}
+
+// Given a chain of delegate service accounts, returns an ID token for the last service account.
+func (f *impersonatedTokenFetcher) fetchIDTokenFromChain(serviceAccounts []string, audience string) (string, error) {
+	if len(serviceAccounts) == 0 {
+		return "", fmt.Errorf("no service accounts provided")
+	}
+
+	// Retrieve an ID token for the last account in the chain.
+	// An access token is required for all other accounts in the delegation chain.
+	idTokenAccount := serviceAccounts[len(serviceAccounts)-1]
+	accessTokenAccounts := serviceAccounts[:len(serviceAccounts)-1]
+
+	var authToken string
+	for i, sa := range accessTokenAccounts {
+		fmt.Printf("Retrieving access token for delegate in position %v\n", i)
+
+		var err error
+		authToken, err = f.fetchAccessToken(sa, authToken, serviceAccounts[i+1:])
+		if err != nil {
+			return "", fmt.Errorf("unable to get access token: %v", err)
+		}
+	}
+
+	fmt.Println("Retrieving impersonated ID token")
+	idToken, err := f.fetchIDToken(idTokenAccount, authToken, audience)
```

Comment:
instead of fetching the access token one by one manually, is it possible to include SAs as delegates in the request? https://cloud.google.com/iam/docs/reference/credentials/rest/v1/projects.serviceAccounts/generateIdToken#request-body